### PR TITLE
[lldb-dap] Provide `declarationLocation` for variables

### PR DESCRIPTION
**Pull request by vogelsgesang**
This commit implements support for the "declaration location" recently added by microsoft/debug-adapter-protocol#494 to the debug adapter protocol.

For the `declarationLocationReference` we need a variable ID similar to the the `variablesReference`. I decided to simply reuse the `variablesReference` here and renamed `Variables::expandable_variables` and friends accordingly. Given that almost all variables have a declaration location, we now assign those variable ids to all variables.

While `declarationLocationReference` effectively supersedes `$__lldb_extensions.declaration`, I did not remove this extension, yet, since I assume that there are some closed-source extensions which rely on it.

I tested this against VS-Code Insiders. However, VS-Code Insiders currently only supports `valueLoctionReference` and not `declarationLocationReference`, yet. Locally, I hence published the declaration locations as value locations, and VS Code Insiders navigated to the expected places. Looking forward to proper VS Code support for `declarationLocationReference`.

### CODE REVIEW

**Comment by vogelsgesang** [comment]
@walter-erquinigo I took a first stab at implementing the DAP proposal. Seemed rather straightforward. I would love to hear if I missed anything.

Also, I am not sure how to implement `valueLocation`, i.e., a the source location referecend by the value. E.g., for a function pointer, this location should point to the source location where the corresponding function was implemented. If you have any idea on how to do so, I would be interested in hearing about it 🙂

**Comment by walter-erquinigo** [comment]
@vogelsgesang , Mr. Bird, I'm swamped this week with work, but I'll look into this before saturday for sure.

**Comment by vogelsgesang** [comment]
No worries. I will have to rework this commit anyway.

In the meantime, a proposal was merged upstream and also already implemented in VS-Code. However, there were still larger changes which will require a complete rewrite of this commit

**Comment by vogelsgesang** [comment]
Ok, this is ready for review now. The upstream protocol changes were merged in the meantime.

I did leave in the `$__lldb_extensions.declaration` although it is superseded by the `declarationLocationReference`. Let me know in case you prefer me to remove it

File: lldb/test/API/tools/lldb-dap/locations/TestDAP_locations.py
Comment (walter-erquinigo) [inline lldb/test/API/tools/lldb-dap/locations/TestDAP_locations.py:15]: does this fail on windows? Otherwise just test this on Windows.

**Comment by walter-erquinigo** [comment]
@vogelsgesang , the `$__lldb_extensions.declaration` is used by the Mojo extension. Once this change gets it, I'll update the Mojo extension to use the new declaration info and remove `$__lldb_extensions.declaration`.

**Comment by walter-erquinigo** [comment]
@vogelsgesang , can you show a screenshot of the variables pane after this change? I want to make sure that simple types, like ints, are not displayed as having children.

**Review by walter-erquinigo — COMMENTED** [review]
This looks good to me, but I'd like @clayborg  to have a second look.

File: lldb/test/API/tools/lldb-dap/locations/TestDAP_locations.py
Comment (vogelsgesang) [reply lldb/test/API/tools/lldb-dap/locations/TestDAP_locations.py:15]: copy-paste error :/

**Comment by vogelsgesang** [comment]
> @vogelsgesang , the $__lldb_extensions.declaration is used by the Mojo extension. Once this change gets it, I'll update the Mojo extension to use the new declaration info and remove $__lldb_extensions.declaration.

Is that extension's source code available? I wonder how you are currently using the declarationLocation in your frontend, and if your extension is based on `lldb-dap`. If so, would it make sense to upstream your UI integration to `lldb-dap`? Or maybe even to VS-Code itself?

**Comment by vogelsgesang** [comment]
> @vogelsgesang , can you show a screenshot of the variables pane after this change? I want to make sure that simple types, like ints, are not displayed as having children.

**With this commit (and also with the changes from #104589)**

image

Current `main`

image

As you can see, there are no differences re the absence / presence of the "expandable chevron" next to the variables.

That being said, I do find it a bit surprising that function pointers apparently return `true` for `MightHaveChildren`

**Comment by walter-erquinigo** [comment]
> > @vogelsgesang , the $__lldb_extensions.declaration is used by the Mojo extension. Once this change gets it, I'll update the Mojo extension to use the new declaration info and remove $__lldb_extensions.declaration.
> 
> Is that extension's source code available? I wonder how you are currently using the declarationLocation in your frontend, and if your extension is based on `lldb-dap`. If so, would it make sense to upstream your UI integration to `lldb-dap`? Or maybe even to VS-Code itself?

The source code is not available and I couldn't include it in the typescript part of the lldb-dap. I'll explain what I do:

The Mojo extension has a listener that captures the DAP traffic. Then, whenever it receives variable information, which includes `$__lldb_extensions.declaration`, it then invokes the Mojo LSP for all the references of a given variable and then it displays the values of that variable in the source code. It looks like this:

Screenshot 2024-08-21 at 10 16 53 PM

Sadly I can't easily open source that because it requires the cooperation of an LSP. If you think that this feature would be useful to you, assuming that you have a preferred LSP, then we can talk about open sourcing this and adding an LSP hook somewhere. With at least 2 LSPs in question, we could design something relatively generic.

**Review by walter-erquinigo — APPROVED** [review]
This LGTM considering that the screenshot looks right and this passes all the tests.

**Comment by vogelsgesang** [comment]
> I'll explain what I do [...]

Thanks for that context! I don't currently have a LSP in mind. Let's see which UI VS-Code will be providing for this (if any). I hope it will not conflict with your existing UI

**Comment by vogelsgesang** [comment]
> This LGTM considering that the screenshot looks right and this passes all the tests.

Thanks! I assume you still want me to wait for @clayborg's review before merging?

**Comment by walter-erquinigo** [comment]
If @clayborg takes too long to review, I'd suggest merging this and then acting on any late feedback. But give him a few days for sure if possible. He's a better reviewer than me.

File: lldb/test/API/tools/lldb-dap/locations/TestDAP_locations.py
Comment (vogelsgesang) [reply lldb/test/API/tools/lldb-dap/locations/TestDAP_locations.py]: @walter-erquinigo based on https://github.com/llvm/llvm-project/pull/105604, it seems that all DAP tests are in fact disabled for Windows. Should I reintroduce the `SkipIfWindows`?

Comment (walter-erquinigo) [reply lldb/test/API/tools/lldb-dap/locations/TestDAP_locations.py]: Oof, this is quite unfortunate. I remember that not long ago there was someone who was trying to make these tests past on window, but it seems that this effort wasn't successful.
And well, yes, in order to have an easier time with buildbots, feel free to reintroduce the `SkipIfWindows` decorator in your tests.

**Comment by vogelsgesang** [comment]
Hi @clayborg,

Are you still planning to review this (as requested by @walter-erquinigo)? Or are you fine with me landing this change without your review?

**Review by clayborg — APPROVED** [review]

**Comment by vogelsgesang** [comment]
The above test failure above is in `lld`, not `lldb`. I don't see how this could have been caused by my change to `lldb-dap`

**Pull request merged**
